### PR TITLE
Week 3 walkthrough 3

### PR DESCRIPTION
**Pull request by mmurray22**
This PR represents a completion of Week 3 Walkthrough Step 3.

### CODE REVIEW

File: portfolio/pom.xml [hunk 1]
```diff
@@ -23,6 +23,12 @@
       <version>4.0.1</version>
       <scope>provided</scope>
     </dependency>
+    
+    <dependency>
+      <groupId>com.google.code.gson</groupId>
+      <artifactId>gson</artifactId>
+      <version>2.8.6</version>
+    </dependency>
```
Comment (ccondit) [inline portfolio/pom.xml]: nit - i think it will be good practice to keep the dependencies ordered alphabetically by groupId, artifactId. that way it will be easier to find exactly what we're looking for...

File: portfolio/src/main/java/com/google/sps/data/MyComments.java [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package com.google.sps.data;
+
+import java.util.ArrayList;
+import java.util.List;
+
+/** Class containing server statistics. */
+public final class MyComments {
```
Comment (ccondit) [inline portfolio/src/main/java/com/google/sps/data/MyComments.java]: Is this class necessary? Can the comments simply be a list in `DataServlet`?

File: portfolio/src/main/java/com/google/sps/servlets/DataServlet.java [hunk 3]
```diff
@@ -14,19 +14,37 @@
...
-/** Servlet that returns some example content. TODO: modify this file to handle comments data */
+/** Servlet that returns some example content. */
 @WebServlet("/data")
 public class DataServlet extends HttpServlet {
 
   @Override
   public void doGet(HttpServletRequest request, HttpServletResponse response) throws IOException {
-    response.setContentType("text/html;");
-    response.getWriter().println("<h1>Hello world Michaela!</h1>");
+    ArrayList<String> comments = new ArrayList<String>();
```
Comment (ccondit) [inline portfolio/src/main/java/com/google/sps/servlets/DataServlet.java]: java can infer the types during generic instance creation:
```suggestion
    ArrayList<String> comments = new ArrayList<>();
```

File: portfolio/src/main/java/com/google/sps/servlets/DataServlet.java [hunk 4]
```diff
@@ -14,19 +14,37 @@
...
+    comments.add("Comment 1: This blog is interesting!");
+    comments.add("Comment 2: Well known information. Not too interesting.");
+    comments.add("Comment 3: Needs some more development, but cool concept!");
+    MyComments myComments = new MyComments(comments);
+    String comment_json = convertToJson(myComments);
```
Comment (ccondit) [inline portfolio/src/main/java/com/google/sps/servlets/DataServlet.java]: style nit - [local vars should be lowerCamelCase](https://google.github.io/styleguide/javaguide.html#s5.2.7-local-variable-names).

File: portfolio/src/main/java/com/google/sps/servlets/DataServlet.java [hunk 5]
```diff
@@ -14,19 +14,37 @@
...
+
+    //Send JSON as the response
+    response.setContentType("application/json;");
+    response.getWriter().println(comment_json);
+  }
+
+  private String convertToJson(MyComments myComments) {
```
Comment (ccondit) [inline portfolio/src/main/java/com/google/sps/servlets/DataServlet.java]:
```suggestion
  private static String convertToJson(MyComments myComments) {
```

File: portfolio/src/main/webapp/script.js [hunk 6]
```diff
@@ -70,3 +70,15 @@ function contact() {
 		</ul>`; 
 	console.log('Contact!'); 
 }
+
+function getComments() {
+    console.log('Hello!');
```
Comment (ccondit) [inline portfolio/src/main/webapp/script.js]: nit - match the indentation here

**Review by ccondit — CHANGES_REQUESTED** [review]

File: portfolio/pom.xml (hunk 1 above)
Comment (mmurray22) [reply portfolio/pom.xml]: thanks! updated!

File: portfolio/src/main/java/com/google/sps/servlets/DataServlet.java (hunk 3 above)
Comment (mmurray22) [reply portfolio/src/main/java/com/google/sps/servlets/DataServlet.java]: ah cool! thanks!

File: portfolio/src/main/java/com/google/sps/servlets/DataServlet.java (hunk 4 above)
Comment (mmurray22) [reply portfolio/src/main/java/com/google/sps/servlets/DataServlet.java]: oh yeah-thanks! updated!

File: portfolio/src/main/java/com/google/sps/servlets/DataServlet.java (hunk 5 above)
Comment (mmurray22) [reply portfolio/src/main/java/com/google/sps/servlets/DataServlet.java]: thanks! updated!

File: portfolio/src/main/webapp/script.js (hunk 6 above)
Comment (mmurray22) [reply portfolio/src/main/webapp/script.js]: hm I don't see any odd indentations on my side...I will push and then look at it in the github UI!

File: portfolio/pom.xml [hunk 7]
```diff
@@ -35,10 +41,10 @@
         <version>2.2.0</version>
         <configuration>
           <!-- TODO: set project ID. -->
-          <deploy.projectId>YOUR_PROJECT_ID_HERE</deploy.projectId>
+          <deploy.projectId>michaelamurray-step-2020</deploy.projectId>
```
Comment (ccondit) [inline portfolio/pom.xml]: nit - leave this for a separate PR.

File: portfolio/src/main/java/com/google/sps/data/MyComments.java (hunk 2 above)
Comment (ccondit) [reply portfolio/src/main/java/com/google/sps/data/MyComments.java]: This is looking good - but still curious if this class is required...

**Review by ccondit — CHANGES_REQUESTED** [review]

**Comment by ccondit** [comment]
I might have triggered a review too early here - looks like there a still a couple of outstanding comments:

- https://github.com/mmurray22/my-portfolio/pull/14#discussion_r447916233
- https://github.com/mmurray22/my-portfolio/pull/14#discussion_r447916515

File: portfolio/pom.xml (hunk 7 above)
Comment (mmurray22) [reply portfolio/pom.xml]: ok-noted!

Comment (mmurray22) [reply portfolio/pom.xml]: updated!

Comment (ccondit) [reply portfolio/pom.xml]: Looks like the change to the `projectId` is still in this PR? Can we remove this s.t. the only change in the POM is adding `gson` as a dependency?

File: portfolio/src/main/java/com/google/sps/data/MyComments.java (hunk 2 above)
Comment (ccondit) [reply portfolio/src/main/java/com/google/sps/data/MyComments.java]: See my previous comment - can the comments simply be a list in `DataServlet`?

**Review by ccondit — CHANGES_REQUESTED** [review]

File: portfolio/README.md [hunk 8]
```diff
@@ -5,4 +5,4 @@ command:
 
 ```bash
 mvn package appengine:run
-```
\ No newline at end of file
+```
```
Comment (ccondit) [inline portfolio/README.md:8]: nit - revert this change since it's unrelated to the PR.

File: portfolio/src/main/java/com/google/sps/servlets/DataServlet.java [hunk 9]
```diff
@@ -15,18 +15,34 @@
...
-/** Servlet that returns some example content. TODO: modify this file to handle comments data */
+/** Servlet that returns some example content. */
 @WebServlet("/data")
 public class DataServlet extends HttpServlet {
 
   @Override
   public void doGet(HttpServletRequest request, HttpServletResponse response) throws IOException {
-    response.setContentType("text/html;");
-    response.getWriter().println("<h1>Hello world Michaela!</h1>");
+    ArrayList<String> comments = new ArrayList<>();
```
Comment (ccondit) [inline portfolio/src/main/java/com/google/sps/servlets/DataServlet.java]: Couple of comments here - one actionable and one for future thought:

1. Per [effective java item 64](go/ej3e-64) (if that go link doesn't resolve I believe i shared the PDF with the pod previously): refer to objects by their interfaces:
```suggestion
    List<String> comments = new ArrayList<>();
```
2. Have a gander at [Guava's ImmutableCollections](https://github.com/google/guava/wiki/ImmutableCollectionsExplained) - no need to change anything now, but these types (and guava, generally) will probably be useful for the capstone project.

File: portfolio/src/main/webapp/script.js [hunk 10]
```diff
@@ -70,3 +70,14 @@ function contact() {
...
 }
+
+function getComments() {
+	fetch('/data').then(response => response.json()).then((comments) => {
+		console.log(comments);
+		const commentElement = document.body;
+		commentElement.innerHTML += 'Here are the comments: \n';
+        for (var i = 0; i < comments.length; i++) {
+            commentElement.innerHTML += comments[i] + '\n';
+        }
```
Comment (ccondit) [inline portfolio/src/main/webapp/script.js]: nit - Reading over[ the style guide](https://google.github.io/styleguide/jsguide.html#features-for-loops) looks like `for-of` loops should be preferred (and are also a bit more readable).

File: portfolio/src/main/java/com/google/sps/servlets/DataServlet.java [hunk 11]
```diff
@@ -15,18 +15,34 @@
...
+    comments.add("Comment 2: Well known information. Not too interesting.");
+    comments.add("Comment 3: Needs some more development, but cool concept!");
+    String commentJSON = convertToJson(comments);
+
+    //Send JSON as the response
+    response.setContentType("application/json;");
+    response.getWriter().println(commentJSON);
+  }
+
+  private static String convertToJson(ArrayList<String> comments) {
```
Comment (ccondit) [inline portfolio/src/main/java/com/google/sps/servlets/DataServlet.java]:
```suggestion
  private static String convertToJson(List<String> comments) {
```

File: portfolio/src/main/webapp/script.js [hunk 12]
```diff
@@ -70,3 +70,14 @@ function contact() {
 		</ul>`; 
 	console.log('Contact!'); 
 }
+
+function getComments() {
+	fetch('/data').then(response => response.json()).then((comments) => {
+		console.log(comments);
+		const commentElement = document.body;
```
Comment (ccondit) [inline portfolio/src/main/webapp/script.js]: nit - this variable name might be misleading since it's actually the `<body>` element.

**Review by ccondit — CHANGES_REQUESTED** [review]

File: portfolio/README.md (hunk 8 above)
Comment (mmurray22) [reply portfolio/README.md:8]: thanks! updated!

File: portfolio/src/main/webapp/script.js (hunk 12 above)
Comment (mmurray22) [reply portfolio/src/main/webapp/script.js]: thanks! updated!

File: portfolio/src/main/webapp/script.js (hunk 10 above)
Comment (mmurray22) [reply portfolio/src/main/webapp/script.js]: thanks! updated!

**Review by ccondit — APPROVED** [review]
Cool! \o/

File: portfolio/src/main/java/com/google/sps/servlets/DataServlet.java [hunk 13]
```diff
@@ -15,18 +15,34 @@
...
   public void doGet(HttpServletRequest request, HttpServletResponse response) throws IOException {
-    response.setContentType("text/html;");
-    response.getWriter().println("<h1>Hello world Michaela!</h1>");
+    List<String> comments = new ArrayList<>();
+    comments.add("Comment 1: This blog is interesting!");
+    comments.add("Comment 2: Well known information. Not too interesting.");
+    comments.add("Comment 3: Needs some more development, but cool concept!");
+    String commentJSON = convertToJson(comments);
+
+    //Send JSON as the response
```
Comment (kaveh096) [inline portfolio/src/main/java/com/google/sps/servlets/DataServlet.java]:
```suggestion
    // Send JSON as the response
```

File: portfolio/src/main/webapp/script.js [hunk 14]
```diff
@@ -70,3 +70,14 @@ function contact() {
 		</ul>`; 
 	console.log('Contact!'); 
 }
+
+function getComments() {
+	fetch('/data').then(response => response.json()).then((comments) => {
+		console.log(comments);
+		const bodyElement = document.body;
+		bodyElement.innerHTML += 'Here are the comments: \n';
```
Comment (kaveh096) [inline portfolio/src/main/webapp/script.js:78]: No action needed. I personally don't find it useful to add to the bottom of the body. I prefer having a dedicated div and add to that so i can control the looks of the page with html and not java server. but it's up to you.

**Review by kaveh096 — CHANGES_REQUESTED** [review]

File: portfolio/src/main/webapp/script.js (hunk 14 above)
Comment (mmurray22) [reply portfolio/src/main/webapp/script.js:78]: Yeah, I think I was just doing that for now to keep the non-serve-let related changes to a minimum while I went through the exercises :)

**Review by kaveh096 — APPROVED** [review]
